src/lib/storage/run_length_segment/run_length_segment_iterable.hpp

```diff
@@ -44,143 +43,192 @@ class RunLengthSegmentIterable : public PointAccessibleSegmentIterable<RunLength
  private:
   const RunLengthSegment<T>& _segment;
 
+  /**
+   * Due to the nature of Run Length encoding, point-access is not in O(1).
+   * Run length segments store the end positions of runs in a sorted vector. To access a particular position, this
+   * vector has to be searched for the end position of the according run. It is possible to find the value of a
+   * position in O(log(n)) by doing a binary search.
+   * In case we have already searched for a nearby position (e.g., when a previous not selective filter removed only
+   * relatively few lines), we can linearly search the end position. Since position lists are often ordered and modern
+   * CPUs are well at prefetching, linear searches are often faster than a complete binary search.
+   * determine_linear_search_offset_distance_threshold() estimates the threshold of when to use a linear or a binary
+   * search. Given the previous and the current chunk offset, the threshold is used to determine if a linear search is
+   * faster for the given number of skipped positions.
+   */
+
+  // LINEAR_SEARCH_ELEMENTS_THRESHOLD denotes the distance for that a linear search is faster than using a binary
+  // search on a sorted vector. The value of 200 has been found by a set of simple TPC-H measurements (#2038).
+  static constexpr auto LINEAR_SEARCH_VECTOR_DISTANCE_THRESHOLD = 200.0f;
+
+  // This method estimates the threshold up to which linear search is faster. Since the end position vector is
+  // a highly compressed vector of offsets (depending on actual data), the offset distance up to which is linear search
+  // is beneficial can be much larger than LINEAR_SEARCH_VECTOR_DISTANCE_THRESHOLD.
+  // Example: for a sorted segment of 65k elements with 10 distinct elements, the expected run length is 6.5k. We would
+  // use linear searches for offset distances of LINEAR_SEARCH_VECTOR_DISTANCE_THRESHOLD * 6.5. For sorted vectors with
+  // few distinct values (superb cases for run length encoding), binary searches are barely used.
+  static ChunkOffset determine_linear_search_offset_distance_threshold(
+      const std::shared_ptr<const pmr_vector<ChunkOffset>>& end_positions) {
+    if (end_positions->empty()) {
+      return 0;
+    }
+
+    const ChunkOffset chunk_size = end_positions->back();
+    const size_t run_count = end_positions->size();
+
+    const auto avg_elements_per_run = static_cast<float>(chunk_size) / run_count;
+    return static_cast<ChunkOffset>(LINEAR_SEARCH_VECTOR_DISTANCE_THRESHOLD * std::ceil(avg_elements_per_run));
+  }
+
+  using EndPositionIterator = typename pmr_vector<ChunkOffset>::const_iterator;
+  static EndPositionIterator search_end_positions_for_chunk_offset(
+      const ChunkOffset previous_chunk_offset, const ChunkOffset chunk_offset,
+      const size_t previous_end_position_index, const size_t linear_search_threshold,
+      const std::shared_ptr<const pmr_vector<ChunkOffset>>& end_positions) {
```

Comment:
I now understand better what is happening here. My suggestion to slightly clean up the interface are:

* move end_positions to the front (no idea why, it just feels better to put the thing you operating on first)
* chunk_offset is confusing here. Isn't it a chunk_offset_offset? How do you feel about `distance`?

Reply:
`chunk_offset` is rather the `current_chunk_offset` that is being searched. The `distance` is currently called `step_size`. Do you mean this `step_size`? I would be fine with `distance` here.

Reply:
Ok, misread it. `old_chunk_offset` and `new_`?